app/src/ui/no-existing-remote-branch/no-existing-remote-branch-warning.tsx

```diff
@@ -0,0 +1,72 @@
+import * as React from 'react'
+import { Button } from '../lib/button'
+import { ButtonGroup } from '../lib/button-group'
+import { Dialog, DialogContent, DialogFooter } from '../dialog'
+import { Dispatcher } from '../../ui/dispatcher'
+import { Repository } from '../../models/repository'
+import { Ref } from '../lib/ref'
+
+interface INoExistingRemoteBranchWarningProps {
+  readonly onDismissed: () => void
+  readonly dispatcher: Dispatcher
+  readonly branchName: string
+  readonly repository: Repository
+}
+
+/** A dialog to display a branch that does not exist on remote */
+export class NoExistingRemoteBranchWarning extends React.Component<
+  INoExistingRemoteBranchWarningProps
+> {
+  private closeButton: Button | null = null
+
+  public constructor(props: INoExistingRemoteBranchWarningProps) {
+    super(props)
+  }
+
+  private onCloseButtonRef = (button: Button | null) => {
+    this.closeButton = button
+  }
+
+  public componentDidMount() {
+    // Since focus is given to the publish branch button by default, set focus to the cancel button.
+    if (this.closeButton != null) {
+      this.closeButton.focus()
+    }
+  }
+
+  public render() {
+    return (
+      <Dialog
+        id="no-existing-remote-branch"
+        title={
+          __DARWIN__ ? 'No Existing Remote Branch' : 'No existing remote branch'
+        }
+        onDismissed={this.props.onDismissed}
+        type="warning"
+      >
+        <DialogContent>
+          <p>
+            The branch <Ref>{this.props.branchName}</Ref> does not exist on
+            remote.
+          </p>
+        </DialogContent>
+
+        <DialogFooter>
+          <ButtonGroup destructive={true}>
+            <Button type="submit" ref={this.onCloseButtonRef}>
+              Cancel
+            </Button>
+            <Button onClick={this.publishBranch}>
+              {__DARWIN__ ? 'Publish Branch' : 'Publish branch'}
+            </Button>
+          </ButtonGroup>
+        </DialogFooter>
+      </Dialog>
+    )
+  }
+
+  private publishBranch = async () => {
+    this.props.dispatcher.closePopup()
+    await this.props.dispatcher.push(this.props.repository)
```

Comment:
@say25 There's some extra detail in here that I don't think we've discussed, and I wanted to write it up and think about it some more before we merge this change.

From the original issue:

> However, I did a pull on the branch while still in the same status, and I got this error:

The root cause of this issue is that the upstream branch no longer exists on the remote, and while `push`-ing to restore the branch on the remote is one option, **what if the upstream branch was removed for a reason?**

Keep in mind that the user is trying to get remote changes into their branch, and if the remote branch doesn't exist any more maybe this is a hint that this workflow should be questioned.

There's another option here, which might not be known to most Git users: removing the tracking information from a branch without removing the local branch from the repository. In Git terminology this is `git branch --unset-upstream`.

If you unset the upstream branch the push/pull button will then render `Publish branch` to the user, which is a potentially more explicit step.

I'm going to think on this some more and report back, I just wanted to register my concern about only having the option to push the branch to the remote.